### PR TITLE
[WIP] Move AssemblyName to managed

### PR DESCRIPTION
**Pull request by kumpera**
The plan is to move all usage of AssemblyName in native to managed.

This PR includes one out of three icalls. Stay tunned for more patches!

### CODE REVIEW

File: mcs/class/corlib/System.Reflection/AssemblyName.cs [hunk 1]
```diff
@@ -84,17 +85,30 @@ public AssemblyName ()
...
 			if (assemblyName == null)
 				throw new ArgumentNullException ("assemblyName");
 			if (assemblyName.Length < 1)
 				throw new ArgumentException ("assemblyName cannot have zero length.");
-				
-			if (!ParseName (this, assemblyName))
-				throw new FileLoadException ("The assembly name is invalid.");
+
+
+			MonoAssemblyName nativeName = default (MonoAssemblyName);
```
Comment (marek-safar) [inline mcs/class/corlib/System.Reflection/AssemblyName.cs]: You don't need default it should be fully assigned and I'd move it just before ParseAssemblyName

File: mcs/class/corlib/System.Reflection/AssemblyName.cs [hunk 2]
```diff
@@ -84,17 +85,30 @@ public AssemblyName ()
...
+			bool isVersionDefined, isTokenDefined;
```
Comment (marek-safar) [inline mcs/class/corlib/System.Reflection/AssemblyName.cs]: Same here, move it inside using

File: mcs/class/corlib/Mono/RuntimeMarshal.cs [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+		internal static string PtrToUtf8String (IntPtr ptr)
+		{
+			unsafe {
+				return new String ((sbyte*)ptr);
+			}
+		}
+
+		internal static StringMarshalHelper MarshalString (string str)
+		{
+			return new StringMarshalHelper (str);
```
Comment (marek-safar) [inline mcs/class/corlib/Mono/RuntimeMarshal.cs]: Missing ArgumentNullException check for `str` value

File: mcs/class/corlib/Mono/RuntimeMarshal.cs [hunk 4]
```diff
@@ -0,0 +1,96 @@
+using System;
+using System.Runtime.InteropServices;
+using System.Runtime.CompilerServices;
+
+namespace Mono {
+	internal struct StringMarshalHelper : IDisposable {
```
Comment (marek-safar) [inline mcs/class/corlib/Mono/RuntimeMarshal.cs]: I'd call it StringToMonoString

**Comment by lambdageek** [comment]
Merged a version of this (#3335)

**Pull request closed without merging**